// neovim/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadPlugin = exports.NvimPlugin = exports.Command = exports.Autocmd = exports.Function = exports.Plugin = exports.Window = exports.Tabpage = exports.Buffer = exports.NeovimClient = exports.Neovim = exports.attach = void 0;
var attach_1 = require("./attach");
Object.defineProperty(exports, "attach", { enumerable: true, get: function () { return attach_1.attach; } });
var index_1 = require("./api/index");
Object.defineProperty(exports, "Neovim", { enumerable: true, get: function () { return index_1.Neovim; } });
Object.defineProperty(exports, "NeovimClient", { enumerable: true, get: function () { return index_1.NeovimClient; } });
Object.defineProperty(exports, "Buffer", { enumerable: true, get: function () { return index_1.Buffer; } });
Object.defineProperty(exports, "Tabpage", { enumerable: true, get: function () { return index_1.Tabpage; } });
Object.defineProperty(exports, "Window", { enumerable: true, get: function () { return index_1.Window; } });
var plugin_1 = require("./plugin");
Object.defineProperty(exports, "Plugin", { enumerable: true, get: function () { return plugin_1.Plugin; } });
Object.defineProperty(exports, "Function", { enumerable: true, get: function () { return plugin_1.Function; } });
Object.defineProperty(exports, "Autocmd", { enumerable: true, get: function () { return plugin_1.Autocmd; } });
Object.defineProperty(exports, "Command", { enumerable: true, get: function () { return plugin_1.Command; } });
var NvimPlugin_1 = require("./host/NvimPlugin");
Object.defineProperty(exports, "NvimPlugin", { enumerable: true, get: function () { return NvimPlugin_1.NvimPlugin; } });
var factory_1 = require("./host/factory");
Object.defineProperty(exports, "loadPlugin", { enumerable: true, get: function () { return factory_1.loadPlugin; } });
